harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -28,56 +31,199 @@
+            return Epoch(epochs)
+
+    def _divides(self, steps: int):
+        assert self.value > 0, "TrainUnit value must be > 0"
+        if steps == 0:
+            return False
+        return steps % self.value == 0
+
+
+class Epoch(TrainUnit):
+    pass
+
+
+class Batch(TrainUnit):
+    pass
+
+
+class Record(TrainUnit):
+    pass
+
+
+class _TrainStepType(enum.Enum):
+    CHECKPOINT = "CHECKPOINT"
+    REPORT = "REPORT"
+    VALIDATE = "VALIDATE"
+    TRAIN = "TRAIN"
+
+
+# XXX: should this be private? will be returned from .train() but is only used internally
+class _TrainStep:
+    def __init__(self, step_type: _TrainStepType, unit: TrainUnit):
+        self.step_type = step_type
+        self.unit = unit
+        self.limit_reached = False
+
+
+class ShouldExit(Exception):
+    """
+    ShouldExit breaks out of the top-level train loop from inside function calls.
+    """
+
+    def __init__(self, skip_exit_checkpoint: bool = False):
+        self.skip_exit_checkpoint = skip_exit_checkpoint
+
+
+class _TrialState:
+    def __init__(
+        self,
+        trial_id: int = 0,
+        last_ckpt: int = 0,
+        steps_completed: int = 0,
+        step_id: int = 0,
+        last_val: int = 0,
+        batches_trained: int = 0,
+        epochs_trained: int = 0,
+    ) -> None:
+        # Store TrialID to distinguish between e.g. pause/restart and continue training.
+        self.trial_id = trial_id
+        self.last_ckpt = last_ckpt
+        self.steps_completed = steps_completed
+        self.step_id = step_id
+        self.last_val = last_val
+        self.batches_trained = batches_trained
+        self.epochs_trained = epochs_trained
+
+
+class PyTorchTrialController:
+    def __init__(
+        self,
+        trial_inst: det.Trial,
+        context: pytorch.PyTorchTrialContext,
+        min_checkpoint_period: TrainUnit,
+        min_validation_period: TrainUnit,
+        average_training_metrics: Optional[bool] = True,
+        smaller_is_better: Optional[bool] = True,
+        max_length: Optional[TrainUnit] = None,
+        steps_completed: Optional[int] = 0,
+        latest_checkpoint: Optional[str] = None,
+        local_training: Optional[bool] = False,
+        test_mode: Optional[bool] = False,
+        # XXX: should default be maxsize or 100?
+        scheduling_unit: Optional[int] = sys.maxsize,
+        det_profiler: Optional[ProfilerAgent] = DummyProfilerAgent(),
+        searcher_metric_name: Optional[str] = None,
+        debug: Optional[bool] = False,
+        checkpoint_policy: Optional[str] = "best",
+        step_zero_validation: Optional[bool] = False,
+    ) -> None:
 
         if not isinstance(trial_inst, PyTorchTrial):
             raise TypeError("PyTorchTrialController requires a PyTorchTrial.")
-        self.trial = trial_inst
-        self.context = cast(pytorch.PyTorchTrialContext, self.context)
-        self.context._set_determined_profiler(self.prof)
+        self._trial = trial_inst
```

Comment:
is PyTorchTrialController going to be user-facing?

It looks like all the symbols are `_`-prefixed on it.  Which isn't the end of the world if it is an internal object.  I care a lot less about internal object attribute names.  I'm only curious.

Reply:
i kept all the objects private because i figured we might want to expose `PyTorchTrialController` at some point in the future. i can remove them if they're annoying.

Reply:
I'd remove them.  But I'd make `PyTorchTrialController` itself private, I think.